File: python/allwright/_web_locators.py
"""Web locator construction; evaluation is owned by the web surface."""
from __future__ import annotations
import json
import re
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from ._locator import Locator

TextMatcher = str | re.Pattern[str]

def semantic_selector(spec: dict) -> str:
    def encode(value):
        if isinstance(value, re.Pattern):
            unsupported = value.flags & ~(re.I | re.M | re.S | re.U)
            if unsupported:
                raise ValueError('Unsupported locator regular expression flags')
            return {'regex': value.pattern, 'flags': ''.join(flag for mask, flag in [(re.I, 'i'), (re.M, 'm'), (re.S, 's')] if value.flags & mask)}
        raise TypeError(f'Unsupported selector value: {type(value)}')
    return 'aw=' + json.dumps(json.dumps(spec, default=encode))

class WebLocators:
    def locator(self, selector: str) -> Locator:
        raise NotImplementedError

    def get_by_role(self, role: str, *, name: TextMatcher | None = None, exact: bool = False,
                    checked: bool | None = None, disabled: bool | None = None,
                    expanded: bool | None = None, include_hidden: bool = False,
                    level: int | None = None, pressed: bool | None = None,
                    selected: bool | None = None) -> Locator:
        if level is not None and (isinstance(level, bool) or not isinstance(level, int) or level < 1):
            raise ValueError('Role level must be a positive integer')
        spec = dict(kind='role', role=role, name=name, exact=exact, checked=checked,
                    disabled=disabled, expanded=expanded, includeHidden=include_hidden,
                    level=level, pressed=pressed, selected=selected)
        return self.locator(semantic_selector({k: v for k, v in spec.items() if v is not None}))

    def get_by_text(self, text: TextMatcher, *, exact: bool = False) -> Locator:
        return self.locator(semantic_selector(dict(kind='text', text=text, exact=exact)))

    def get_by_label(self, text: TextMatcher, *, exact: bool = False) -> Locator:
        return self.locator(semantic_selector(dict(kind='label', text=text, exact=exact)))

    def get_by_placeholder(self, text: TextMatcher, *, exact: bool = False) -> Locator:
        return self.locator(semantic_selector(dict(kind='placeholder', text=text, exact=exact)))

    def get_by_alt_text(self, text: TextMatcher, *, exact: bool = False) -> Locator:
        return self.locator(semantic_selector(dict(kind='altText', text=text, exact=exact)))

    def get_by_title(self, text: TextMatcher, *, exact: bool = False) -> Locator:
        return self.locator(semantic_selector(dict(kind='title', text=text, exact=exact)))

    def get_by_test_id(self, text: TextMatcher) -> Locator:
        return self.locator(semantic_selector(dict(kind='testId', text=text)))

File: java/src/main/java/dev/allwright/client/WebLocators.java
package dev.allwright.client;
/** Shared web locator builders for pages and nested locators. Text accepts String or Pattern. */
public interface WebLocators {
    Locator locator(String selector);
    default Locator getByRole(String role) { return getByRole(role, new RoleOptions()); }
    default Locator getByRole(String role, RoleOptions options) { return locator(WebSelectorSupport.role(role, options)); }
    default Locator getByText(Object text) { return getByText(text, new TextOptions()); }
    default Locator getByText(Object text, TextOptions options) { return locator(WebSelectorSupport.text("text", text, options)); }
    default Locator getByLabel(Object text) { return getByLabel(text, new TextOptions()); }
    default Locator getByLabel(Object text, TextOptions options) { return locator(WebSelectorSupport.text("label", text, options)); }
    default Locator getByPlaceholder(Object text) { return getByPlaceholder(text, new TextOptions()); }
    default Locator getByPlaceholder(Object text, TextOptions options) { return locator(WebSelectorSupport.text("placeholder", text, options)); }
    default Locator getByAltText(Object text) { return getByAltText(text, new TextOptions()); }
    default Locator getByAltText(Object text, TextOptions options) { return locator(WebSelectorSupport.text("altText", text, options)); }
    default Locator getByTitle(Object text) { return getByTitle(text, new TextOptions()); }
    default Locator getByTitle(Object text, TextOptions options) { return locator(WebSelectorSupport.text("title", text, options)); }
    default Locator getByTestId(Object text) { return getByTestId(text, new TextOptions()); }
    default Locator getByTestId(Object text, TextOptions options) { return locator(WebSelectorSupport.text("testId", text, options)); }
}
